### PR TITLE
Introduce APIs for jvm modifiers

### PR DESCRIPTION
**Pull request by ting-yuan**
Specifically,
```
ResolverImpl.isJvmTransient(property: KSPropertyDeclaration)
ResolverImpl.isJvmVolatile(property: KSPropertyDeclaration)
ResolverImpl.isJvmStrictfp(function: KSFunctionDeclaration)
ResolverImpl.isJvmSynchronized(function: KSFunctionDeclaration)
```

Those modifiers / access flags are part of Kotlin compilers implementation on the JVM
platform. They don't exist directly in the Kotlin language. Therefore,
those APIs are introduced so that processors don't need to guess by
themselves.

### CODE REVIEW

**Comment by ting-yuan** [comment]
+@ZacSweers, this is a tentative solution.

**Comment by ting-yuan** [comment]
Maybe we can use an unified `ResolverImpl.effectiveJavaModifiers`. Will update later.

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt [hunk 1]
```diff
@@ -1089,6 +1095,82 @@ class ResolverImpl(
...
+            Origin.JAVA -> {
+                addVisibilityModifiers()
+                if (declaration is KSClassDeclaration && declaration.classKind == ClassKind.INTERFACE)
+                    modifiers.add(Modifier.ABSTRACT)
+            }
+            Origin.KOTLIN -> {
+                addVisibilityModifiers()
+                if (!declaration.isOpen())
+                    modifiers.add(Modifier.FINAL)
+                if (declaration.hasAnnotation("kotlin.jvm.JvmStatic"))
```
Comment (neetopia) [inline compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt:1123]: minor: maybe extract consts.

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt [hunk 2]
```diff
@@ -1089,6 +1095,82 @@ class ResolverImpl(
...
+                if (declaration.hasAnnotation("kotlin.jvm.Transient"))
+                    modifiers.add(Modifier.JAVA_TRANSIENT)
+                if (declaration.hasAnnotation("kotlin.jvm.Volatile"))
+                    modifiers.add(Modifier.JAVA_VOLATILE)
+                when (declaration) {
+                    is KSClassDeclaration -> {
+                        if (declaration.isCompanionObject)
+                            modifiers.add(Modifier.JAVA_STATIC)
+                        if (declaration.classKind == ClassKind.INTERFACE)
+                            modifiers.add(Modifier.ABSTRACT)
```
Comment (neetopia) [inline compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt:1142]: Nit: feeling kind of tricky for this one. You can actually write `abstract interface` but it does not have any effects.

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt [hunk 3]
```diff
@@ -734,3 +782,42 @@ internal val KSDeclarationContainer.declarationsInSourceOrder: Sequence<KSDeclar
...
     }
+
+internal val KSPropertyDeclaration.jvmAccessFlag: Int
+    get() = when (origin) {
+        Origin.KOTLIN_LIB -> {
+            val descriptor = (this as KSPropertyDeclarationDescriptorImpl).descriptor
+            val kotlinBinaryJavaClass = descriptor.getContainingKotlinJvmBinaryClass()
+            kotlinBinaryJavaClass?.let {
+                BinaryClassInfoCache.getCached(it).fieldAccFlags.get(this.simpleName.asString()) ?: 0
+            } ?: 0
```
Comment (neetopia) [inline compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt:793]: not sure if return 0 is best here, I feel if the containing class can't be obtained, it rather implies an error?

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt (hunk 2 above)
Comment (ting-yuan) [reply compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt:1142]: This follows JVM [spec](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.1). Java interfaces have `abstract` **effectively**.

> If the ACC_INTERFACE flag is set, the ACC_ABSTRACT flag must also be set, and the ACC_FINAL, ACC_SUPER, and ACC_ENUM flags set must not be set.

Comment (neetopia) [reply compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt:1142]: That's what I mean, interface is `abstract` by itself, adding a modifier seems a little bit unnecessary to me. Either way works for me.

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt (hunk 3 above)
Comment (ting-yuan) [reply compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt:793]: `getContainingKotlinJvmBinaryClass()` has null checks everywhere and I'm not so confident with it. For example, I got a null for some `<init>` of `Companion`. Also there can be some builtin functions, e.g., `Boolean.equals` that doesn't have sources.

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt (hunk 1 above)
Comment (ting-yuan) [reply compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt]: Done.

File: compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt (hunk 3 above)
Comment (ting-yuan) [reply compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt:794]: Added a few comments in code.

**Review by neetopia — APPROVED** [review]

**Pull request merged**